File: abc/abc377/c/main.go
package main

import (
    "bufio"
    "fmt"
    "os"
)

type Point struct {
    x, y int64
}

func main() {
    scanner := bufio.NewScanner(os.Stdin)
    scanner.Scan()
    var n, m int64
    fmt.Sscanf(scanner.Text(), "%d %d", &n, &m)

    knights := make([]Point, m)
    for i := int64(0); i < m; i++ {
        scanner.Scan()
        var x, y int64
        fmt.Sscanf(scanner.Text(), "%d %d", &x, &y)
        knights[i] = Point{x, y}
    }

    attacked := make(map[Point]bool)
    occupied := make(map[Point]bool)

    for _, k := range knights {
        occupied[k] = true
    }

    moves := []Point{
        {2, 1}, {1, 2}, {-1, 2}, {-2, 1},
        {-2, -1}, {-1, -2}, {1, -2}, {2, -1},
    }

    for _, k := range knights {
        for _, move := range moves {
            nx := k.x + move.x
            ny := k.y + move.y
            if nx >= 1 && nx <= n && ny >= 1 && ny <= n {
                attacked[Point{nx, ny}] = true
            }
        }
    }

    for pos := range occupied {
        delete(attacked, pos)
    }

    total := n * n
    unsafeCount := int64(len(attacked) + len(occupied))
    safeCount := total - unsafeCount

    fmt.Println(safeCount)
}

File: abc/abc377/c/main.rs
use std::collections::HashSet;
use std::io::{self, BufRead};

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let first_line = lines.next().unwrap().unwrap();
    let mut iter = first_line.split_whitespace();
    let n: i64 = iter.next().unwrap().parse().unwrap();
    let m: usize = iter.next().unwrap().parse().unwrap();
    
    let mut knights: Vec<(i64, i64)> = Vec::with_capacity(m);
    for _ in 0..m {
        let line = lines.next().unwrap().unwrap();
        let mut iter = line.split_whitespace();
        let x: i64 = iter.next().unwrap().parse().unwrap();
        let y: i64 = iter.next().unwrap().parse().unwrap();
        knights.push((x, y));
    }
    
    let mut attacked = HashSet::new();
    
    let mut occupied = HashSet::new();
    for &(x, y) in &knights {
        occupied.insert((x, y));
    }
    
    let moves = [
        (2, 1), (1, 2), (-1, 2), (-2, 1),
        (-2, -1), (-1, -2), (1, -2), (2, -1)
    ];
    
    for &(x, y) in &knights {
        for &(dx, dy) in &moves {
            let nx = x + dx;
            let ny = y + dy;
            if nx >= 1 && nx <= n && ny >= 1 && ny <= n {
                attacked.insert((nx, ny));
            }
        }
    }
    
    for pos in &occupied {
        attacked.remove(pos);
    }
    
    let total = n * n;
    let unsafe_count = (attacked.len() + occupied.len()) as i64;
    let safe_count = total - unsafe_count;
    
    println!("{}", safe_count);
}